src/server/active_tran_server.cpp

```diff
@@ -104,20 +105,52 @@ active_tran_server::connection_handler::disconnect ()
 void
 active_tran_server::connection_handler::receive_saved_lsa (page_server_conn_t::sequenced_payload &a_ip)
 {
+  auto &node_vec = dynamic_cast<active_tran_server *> (&m_ts)->m_node_vec; // casting to access m_node_vec
+  const auto total_node_cnt = node_vec.size ();
+  const auto quorum = total_node_cnt / 2 + 1; // For now, it's fixed to the number of the majority.
+  std::vector<log_lsa> collected_saved_lsa;
   std::string message = a_ip.pull_payload ();
   log_lsa saved_lsa;
 
   assert (sizeof (log_lsa) == message.size ());
   std::memcpy (&saved_lsa, message.c_str (), sizeof (log_lsa));
 
-  if (log_Gl.m_max_ps_flushed_lsa < saved_lsa)
+  assert (saved_lsa > m_node.get_saved_lsa ()); // increasing monotonically
+  m_node.set_saved_lsa (saved_lsa);
+
+  /*
+   * Gather all PS'es saved_lsa and sort it in ascending order.
+   * The <total_node_count - quorum>'th element is the consensus LSA, upon which the majority (quorumn) of PS agrees.
+   * [5, 5, 6, 9, 10] -> "6" is the consensus LSA.
+   * [9, 10] -> "9"
+   */
+  for (const auto &node : node_vec)
+    {
+      collected_saved_lsa.emplace_back (node->get_saved_lsa ());
+    }
+  std::sort (collected_saved_lsa.begin (), collected_saved_lsa.end ());
+
+  const log_lsa consensus_lsa = collected_saved_lsa[total_node_cnt - quorum];
+  if (log_Gl.get_ps_consensus_flushed_lsa () < consensus_lsa)
     {
-      log_Gl.update_max_ps_flushed_lsa (saved_lsa);
+      log_Gl.update_ps_consensus_flushed_lsa (consensus_lsa);
     }
 
   if (prm_get_bool_value (PRM_ID_ER_LOG_COMMIT_CONFIRM))
     {
-      _er_log_debug (ARG_FILE_LINE, "[COMMIT CONFIRM] Received LSA = %lld|%d.\n", LSA_AS_ARGS (&saved_lsa));
+      std::stringstream ss;
+      const log_lsa consensus_lsa = log_Gl.get_ps_consensus_flushed_lsa ();
+
+      ss << "[COMMIT CONFIRM] Received saved LSA = " << saved_lsa.pageid << "|" << saved_lsa.offset;
```

Comment:
Nit.
I would really use a stack fixed buffer + `snprintf`; it is probably faster; and this is not debug-only code; I know iostream is "fancier" but ... :)

Reply:
done :)